### PR TITLE
Check run exists

### PR DESCRIPTION
**Pull request by janvanrijn**

### CODE REVIEW

File: openml/runs/functions.py [hunk 1]
```diff
@@ -42,6 +44,19 @@ def run_task(task, model):
...
+    flow_id = flow_exists(flow.name, flow.external_version)
+
+    # skips the run if it already exists and the user opts for this in the config file.
+    # also, if the flow is not present on the server, the check is not needed.
+    if config.avoid_duplicate_runs and flow_id:
+        flow = get_flow(flow_id)
+        setup_id = setup_exists(flow, model)
+        ids = _run_exists(task.task_id, setup_id)
+        if ids:
+            raise PyOpenMLError("Run already exists in server. Run id(s): %s" %str(ids))
```
Comment (mfeurer) [inline openml/runs/functions.py]: I wonder whether it should rather return the run instead of raising an error.

File: openml/runs/functions.py [hunk 2]
```diff
@@ -55,19 +70,44 @@ def run_task(task, model):
...
+    try:
+        result = list_runs(task=[task_id], setup=[setup_id])
+        if len(result) > 0:
+            return set(result.keys())
+        else:
+            return False
+    except OpenMLServerException as exception:
+        # error code 512 implies no results. This means the run does not exist yet
+        assert(exception.code == 512)
+        return False
```
Comment (mfeurer) [inline openml/runs/functions.py]: Wouldn't an empty set be a more appropriate return value?

File: openml/runs/run.py [hunk 3]
```diff
@@ -182,55 +182,59 @@ def _create_description_xml(self):
         description_xml = xmltodict.unparse(description, pretty=True)
         return description_xml
 
-def _parse_parameters(model, flow):
-    """Extracts all parameter settings from a model in OpenML format.
+    @staticmethod
```
Comment (mfeurer) [inline openml/runs/run.py]: Hm, why did this become a static method? It doesn't access any class variables, right?

File: openml/setups/functions.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+def setup_exists(downloaded_flow, sklearn_model):
+    '''
+    Checks whether a flow / hyperparameter configuration already exists on the server
+
+    Parameter
+    ---------
+
+    downloaded_flow : flow
+        the openml flow object (should be downloaded from server.
+        Otherwise also give flow id parameter)
```
Comment (mfeurer) [inline openml/setups/functions.py]: Not sure what you mean here. How can one also give the flow id parameter?

File: openml/setups/functions.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    sklearn_model : BaseEstimator
+        The base estimator that was used to create the flow. Will
+         be used to extract parameter settings from.
+
+    Returns
+    -------
+    setup_id : int s
```
Comment (mfeurer) [inline openml/setups/functions.py]: Writing `int` is sufficient here. The description is in the line below.

File: openml/setups/functions.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        setup id iff exists, False otherwise
+    '''
+
+    # sadly, this api call relies on a run object
+    openml_param_settings = openml.runs.OpenMLRun._parse_parameters(sklearn_model, downloaded_flow)
+    description = xmltodict.unparse(_to_dict(downloaded_flow.flow_id, openml_param_settings), pretty=True)
+    file_elements = {'description': ('description.arff',description)}
+
+    result = openml._api_calls._perform_api_call('/setup/exists/',
+                                                 file_elements = file_elements)
```
Comment (mfeurer) [inline openml/setups/functions.py]: Could you remove the whitespaces around the equals sign, please?

File: tests/test_flows/test_flow.py [hunk 7]
```diff
@@ -174,24 +176,29 @@ def test_illegal_flow(self):
...
-                                 components=collections.OrderedDict(),
-                                 parameters=collections.OrderedDict(),
-                                 parameters_meta_info=collections.OrderedDict(),
-                                 external_version=_format_external_version(
-                                     'sklearn', sklearn.__version__),
-                                 tags=[],
-                                 language='English',
-                                 dependencies='')
+        flow_id = openml.flows.flow_exists(name, version)
+        self.assertEquals(flow_id, False)
```
Comment (mfeurer) [inline tests/test_flows/test_flow.py]: Could you please use `self.assertFalse` here?

File: tests/test_flows/test_flow.py [hunk 8]
```diff
@@ -174,24 +176,29 @@ def test_illegal_flow(self):
                                                    ('classif', sklearn.tree.DecisionTreeClassifier())])
         self.assertRaises(ValueError, openml.flows.sklearn_to_flow, illegal)
 
-    def test_ensure_flow_exists(self):
-        sentinel = get_sentinel()
+    def test_nonexiting_flow_exists(self):
```
Comment (mfeurer) [inline tests/test_flows/test_flow.py]: There is a typo in the function name.

File: tests/test_flows/test_flow.py [hunk 9]
```diff
@@ -174,24 +176,29 @@ def test_illegal_flow(self):
...
-                                 parameters_meta_info=collections.OrderedDict(),
-                                 external_version=_format_external_version(
-                                     'sklearn', sklearn.__version__),
-                                 tags=[],
-                                 language='English',
-                                 dependencies='')
+        flow_id = openml.flows.flow_exists(name, version)
+        self.assertEquals(flow_id, False)
+
+    def test_exiting_flow_exists(self):
```
Comment (mfeurer) [inline tests/test_flows/test_flow.py]: And the same typo.

File: tests/test_setups/test_setup_functions.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    sentinel = 'TEST%s' % sentinel
+    return sentinel
+
+
+
+class TestRun(TestBase):
+
+    def test_nonexisting_setup_exists(self):
+        from sklearn.tree import DecisionTreeClassifier
+        # first publish a nonexiting flow
```
Comment (mfeurer) [inline tests/test_setups/test_setup_functions.py]: *nonexisting

File: tests/test_setups/test_setup_functions.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    sentinel = md5.hexdigest()[:10]
+    sentinel = 'TEST%s' % sentinel
+    return sentinel
+
+
+
+class TestRun(TestBase):
+
+    def test_nonexisting_setup_exists(self):
+        from sklearn.tree import DecisionTreeClassifier
```
Comment (mfeurer) [inline tests/test_setups/test_setup_functions.py]: Could you import at the beginning of the file?

File: tests/test_setups/test_setup_functions.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        sentinel = get_sentinel()
+        dectree = DecisionTreeClassifier()
+        flow = openml.flows.sklearn_to_flow(dectree)
+        flow.name = 'TEST%s%s' % (sentinel, flow.name)
+        flow.publish()
+
+        # although the flow exists, we can be sure there are no
+        # setups (yet) as it hasn't been ran
+        setup_id = openml.setups.setup_exists(flow, dectree)
+        self.assertEquals(setup_id, False)
```
Comment (mfeurer) [inline tests/test_setups/test_setup_functions.py]: Could you please use `self.assertFalse` here?

File: tests/test_setups/test_setup_functions.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+
+    def test_existing_setup_exists(self):
+        from sklearn.ensemble import BaggingClassifier
```
Comment (mfeurer) [inline tests/test_setups/test_setup_functions.py]: Please import on the top of the file.

File: tests/test_setups/test_setup_functions.py [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                                    max_samples=0.5)
+        flow = openml.flows.sklearn_to_flow(bagging)
+        flow.name = 'TEST%s%s' % (get_sentinel(), flow.name)
+        flow = flow.publish()
+        flow = openml.flows.get_flow(flow.flow_id)
+
+        # although the flow exists, we can be sure there are no
+        # setups (yet) as it hasn't been ran
+        setup_id = openml.setups.setup_exists(flow, bagging)
+        self.assertEquals(setup_id, False)
```
Comment (mfeurer) [inline tests/test_setups/test_setup_functions.py]: Could you please use `self.assertFalse` here?

File: openml/_api_calls.py [hunk 15]
```diff
@@ -80,7 +81,7 @@ def _read_url_files(url, data=None, file_dictionary=None, file_elements=None):
     # 'gzip,deflate'
     response = requests.post(url, data=data, files=file_elements)
     if response.status_code != 200:
-        raise OpenMLServerError(('Status code: %d\n' % response.status_code) + response.text)
+        raise _parse_server_exception(response)
```
Comment (janvanrijn) [inline openml/_api_calls.py]: You are correct. However, from a programming perspective, I would keep it the way it is. "_parse_server_exception" is an understandable function with expected inputs and outputs. The fact that the output is a raisable object, does not mean that the intended behaviour of this function should be raising an error.

File: openml/exceptions.py [hunk 16]
```diff
@@ -2,13 +2,22 @@ class PyOpenMLError(Exception):
     def __init__(self, message):
         super(PyOpenMLError, self).__init__(message)
 
-
+# class for when something is really wrong on the server (result did not parse to dict)
```
Comment (janvanrijn) [inline openml/exceptions.py]: Agreed.

File: openml/exceptions.py [hunk 17]
```diff
@@ -2,13 +2,22 @@ class PyOpenMLError(Exception):
...
 class OpenMLServerError(PyOpenMLError):
-    """Server didn't respond 200."""
+    """Server didn't respond 200, contains unparsed error."""
     def __init__(self, message):
         message = "OpenML Server error: " + message
         super(OpenMLServerError, self).__init__(message)
 
+# class for when the result of the server was not 200 (e.g., listing call w/o results)
```
Comment (janvanrijn) [inline openml/exceptions.py]: Agreed

File: openml/flows/functions.py [hunk 18]
```diff
@@ -69,6 +70,41 @@ def list_flows(offset=None, size=None, tag=None):
     return _list_flows(api_call)
 
 
+def flow_exists(name, version):
+    """Retrieves the flow id of the flow uniquely identified by name+version.
```
Comment (janvanrijn) [inline openml/flows/functions.py]: Fixed the documentation + param name

File: openml/flows/functions.py [hunk 19]
```diff
@@ -69,6 +70,41 @@ def list_flows(offset=None, size=None, tag=None):
...
+
+    xml_response = _perform_api_call("flow/exists",
+                                     data={'name': name, 'external_version': version})
+
+    result_dict = xmltodict.parse(xml_response)
+    flow_id = int(result_dict['oml:flow_exists']['oml:id'])
+    if flow_id > 0:
+        return flow_id
+    else:
+        return False;
```
Comment (janvanrijn) [inline openml/flows/functions.py]: agreed

File: openml/runs/functions.py [hunk 20]
```diff
@@ -42,6 +44,19 @@ def run_task(task, model):
     # TODO move this into its onwn module. While it somehow belongs here, it
     # adds quite a lot of functionality which is better suited in other places!
     # TODO why doesn't this accept a flow as input? - this would make this more flexible!
+    flow = sklearn_to_flow(model)
+
+    # returns flow id if the flow exists on the server, -1 otherwise
```
Comment (janvanrijn) [inline openml/runs/functions.py]: Fixed

File: openml/runs/functions.py [hunk 21]
```diff
@@ -42,6 +44,19 @@ def run_task(task, model):
...
+    flow_id = flow_exists(flow.name, flow.external_version)
+
+    # skips the run if it already exists and the user opts for this in the config file.
```
Comment (janvanrijn) [inline openml/runs/functions.py]: Fine.

File: openml/runs/functions.py (hunk 1 above)
Comment (janvanrijn) [inline openml/runs/functions.py]: Idk, in some sense the user exhibits unwanted behavior, i.e., he explicitly states that he doesn't want to have duplicate runs on the server, yet he didn't (correctly) check  whether the run is already on the server. This should be reported to the user.

Furthermore, as in this case the variables are static, I see no problem in raising the error.

File: openml/runs/functions.py (hunk 2 above)
Comment (janvanrijn) [inline openml/runs/functions.py]: IMO not. It is in principle a boolean function, that potentially gives more information (i.e., if there are runs the non-empty set evaluates to True).

File: openml/runs/run.py (hunk 3 above)
Comment (janvanrijn) [inline openml/runs/run.py]: That's exactly why it became a static method.

File: openml/setups/functions.py (hunk 4 above)
Comment (janvanrijn) [inline openml/setups/functions.py]: Comments were from old version of the code. removed the confusing part

File: openml/setups/functions.py (hunk 5 above)
Comment (janvanrijn) [inline openml/setups/functions.py]: Typo, I suppose. Fixed it

File: openml/setups/functions.py (hunk 6 above)
Comment (janvanrijn) [inline openml/setups/functions.py]: Done

File: tests/test_flows/test_flow.py (hunk 8 above)
Comment (janvanrijn) [inline tests/test_flows/test_flow.py]: Fixed

File: tests/test_flows/test_flow.py (hunk 7 above)
Comment (janvanrijn) [inline tests/test_flows/test_flow.py]: Done

File: tests/test_flows/test_flow.py (hunk 9 above)
Comment (janvanrijn) [inline tests/test_flows/test_flow.py]: Fixed

File: tests/test_setups/test_setup_functions.py (hunk 11 above)
Comment (janvanrijn) [inline tests/test_setups/test_setup_functions.py]: Done

File: tests/test_setups/test_setup_functions.py (hunk 10 above)
Comment (janvanrijn) [inline tests/test_setups/test_setup_functions.py]: Done

File: tests/test_setups/test_setup_functions.py (hunk 12 above)
Comment (janvanrijn) [inline tests/test_setups/test_setup_functions.py]: Done

File: tests/test_setups/test_setup_functions.py (hunk 13 above)
Comment (janvanrijn) [inline tests/test_setups/test_setup_functions.py]: Done

File: openml/_api_calls.py (hunk 15 above)
Comment (mfeurer) [inline openml/_api_calls.py]: k

File: openml/runs/run.py (hunk 3 above)
Comment (mfeurer) [inline openml/runs/run.py]: But it wasn't a method before, so I don't see why it became a method in the first place.

**Pull request merged**